ocs/site_config.py

```diff
@@ -575,4 +592,104 @@ def scan_for_agents(do_registration=True):
     return items
-                    
+
+
+class ArgContainer:
+    """
+    A container to store a list of args as a dictionary, with the argument names
+    (beginning with a hyphen) as keys, and list of arguments as values. Any
+    arguments passed before an argument key is put under the '__positional__'
+    key, even though positional arguments aren't really supported by ocs agents
+    or the site-config....
+
+    Args:
+        args (list):
+            Argument list (each item should be a single word)
+
+    Attributes:
+        arg_dict (dict):
+            Dictionary of arguments, indexed by argument keyword.
+    """
+    def __init__(self, args):
+        self.arg_dict = collections.OrderedDict()
+
+        cur_key = '__positional__'
+        self.arg_dict[cur_key] = []
+        for arg in args:
+            if arg[0] == '-':
+                cur_key = arg
+                self.arg_dict[cur_key] = []
+            else:
+                self.arg_dict[cur_key].append(arg)
+
+    def update(self, arg_container2):
+        """
+        Updates the arg_dict with the arg_dict from another ArgContainer
+
+        Args:
+            arg_container2 (ArgContainer):
+                The other ArgContainer with which you want to update the arg_dict.
+        """
+        self.arg_dict.update(arg_container2.arg_dict)
+
+    def to_list(self):
+        """
+        Returns the argument list representation of this container.
+        """
+        arg_list = []
+        for k, v in self.arg_dict.items():
+            if k is not '__positional__':
+                arg_list.append(k)
+            arg_list.extend(v)
+
+        return arg_list
+
+
+def parse_args(agent_class=None, parser=None):
+    """
+    Function to parse site-config and agent arguments. This function takes
+    site, host, and instance arguments into account by making sure the instance
+    arguments get passed through the arg_parse parser. This helps make sure
+    units and options are consistent with those defined by the argparse
+    argument, even when the arguments come from the site-config file and not
+    the command line.
+
+    Args:
+        agent_class (str, optional):
+            Name of the Agent class.  This
+            may be matched against the agent_class name provided by
+            the Agent instance, as a way of finding the right
+            InstanceConfig.
+        parser (argparse.ArgumentParser, optional):
+            Argument parser containing agent-specific arguments.
+            If None, an empty parser will be created.
+    """
+
+    pre_parser = argparse.ArgumentParser()
+    add_arguments(pre_parser)
+    pre_args, _ = pre_parser.parse_known_args(args=sys.argv[1:])
+
+    site, host, instance = get_config(pre_args, agent_class=agent_class)
+
+    # Container from command line args
+    cl_container = ArgContainer(sys.argv[1:])
+
+    arg_list = []
+    for arg in instance.arguments:
+        arg_list.extend(map(str, arg))
```

Comment:
An interesting consequence of this (all this) is that the site_config yaml arguments don't have to be organized as key-value pairs.  Our current syntax of
```
       'arguments': [['--initial-state', 'record'],
                     ['--data-dir', './data']]
```
is now equivalent to:
```
       'arguments': [['--initial-state', 'record', '--data-dir', './data']]
```
but the following, more sensible looking thing, will cause trouble:
```
       'arguments': ['--initial-state', 'record', '--data-dir', './data']
```
It might be worth allowing this latter case to work too... you can just check for isinstance(arg, str).